File: swift/Tests/AksaraTests/TestSupport.swift
import Foundation
import XCTest
@testable import Aksara

/// A scripted `RemoteBundleFetcher` for OTA tests. Records what was requested and
/// returns queued outcomes in order.
final class MockFetcher: RemoteBundleFetcher, @unchecked Sendable {
    private var outcomes: [Result<FetchOutcome, Error>]
    private(set) var requestedURLs: [URL] = []
    private(set) var sentETags: [String?] = []

    init(_ outcomes: [Result<FetchOutcome, Error>]) {
        self.outcomes = outcomes
    }

    func fetch(url: URL, etag: String?) async throws -> FetchOutcome {
        requestedURLs.append(url)
        sentETags.append(etag)
        guard !outcomes.isEmpty else { throw OTAError.invalidResponse }
        return try outcomes.removeFirst().get()
    }
}

enum TestSupport {
    /// A throwaway directory unique per call — keeps each test's disk cache isolated.
    static func makeTempDir() -> URL {
        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent("AksaraTests-\(UUID().uuidString)", isDirectory: true)
        try? FileManager.default.createDirectory(at: url, withIntermediateDirectories: true)
        return url
    }

    static func data(_ json: String) -> Data { Data(json.utf8) }
}

File: kotlin/aksara-core/src/test/kotlin/com/aksara/TestSupport.kt
package com.aksara

import java.io.File
import kotlin.io.path.createTempDirectory

/**
 * A scripted [RemoteBundleFetcher] for OTA tests. Records what was requested and
 * returns queued outcomes in order. Mirror of the Swift `MockFetcher`.
 */
class MockFetcher(outcomes: List<Result<FetchOutcome>>) : RemoteBundleFetcher {
    private val queue = ArrayDeque(outcomes)
    val requestedUrls = mutableListOf<String>()
    val sentEtags = mutableListOf<String?>()

    override suspend fun fetch(url: String, etag: String?): FetchOutcome {
        requestedUrls += url
        sentEtags += etag
        val next = queue.removeFirstOrNull() ?: throw OtaException("no more outcomes")
        return next.getOrThrow()
    }
}

object TestSupport {
    /** A throwaway directory unique per call — keeps each test's disk cache isolated. */
    fun makeTempDir(): File = createTempDirectory("AksaraTests").toFile()

    /** Loads bundled test JSON (`en.json`, ...) from the test classpath. */
    fun classpathLoader(): (String) -> ByteArray? = { language ->
        TestSupport::class.java.classLoader
            ?.getResourceAsStream("$language.json")
            ?.use { it.readBytes() }
    }

    fun bytes(json: String): ByteArray = json.toByteArray()
}
